docs/expert-tcp-udp.md

```diff
@@ -0,0 +1,118 @@
+
+* M4 arrives, but a router along the way spontaneously duplicated the message in transit, so it arrives *twice*, in second *and *third place. TCP is clever though and knows to discard the duplicate M4.
+
+TCP wants to protect you from this wrong and incomplete order, however. As M2, M3, and M4 arrive, TCP holds those messages until M1 arrives (this is known as *head of line blocking*). While M1 is lost, TCP waits for a while to acknowledge what it has received (delayed ACK) and then waits some more while M1 is resent. Eventually, all four messages are received and TCP hands off the data to your application.
+
+As far as TCP is concerned, this is how it’s supposed to happen. It’s more reliable, but has noticeably more overhead in the form of round-trip exchanges. There are adjustments that can be made to minimize some of the more problematic features (such as turning off delayed ACK), but to get the guarantees of TCP, your application must necessarily submit to these constraints.
+
+Unfortunately, the protocols themselves only represent two points on the reliability gradient. TCP doesn’t have a reduced guarantees mode and UDP doesn’t have an opt-in for more reliability. If our only choices involved choosing TCP or UDP, we’d face a hard decision between the risks of UDP’s best effort and the restrictions of TCP. Luckily, your game has a say in the reliability story, too.
+
+## Optionally reliable
+
+Your game, as a networked application, can choose how to use the network.
+
+In general, your game is not fixed to a single protocol: you can use UDP when you don’t require the reliability guarantees of TCP and you can switch to TCP when you do. Under good network conditions (low latency and packet loss), this won’t matter much. But if things get shaky, then you’ll avoid some misadventures associated with lost messages.
+
+For points on the gradient between optional and required, there’s a strategy called *optionally reliable UDP*. This is when you build additional guarantees in your application, rather than relying on the protocol. In outline, this is sending UDP packets that require some response from the recipient; the message is re-sent until the expected response is received.
+
+But whether you choose UDP or TCP or optionally reliable UDP comes down to the *content* of your messages and what they mean to your game and players.
+
+## Start with your payloads 
+
+The content of your messages matters more than the vehicle they travel in. Start by focusing on your payloads, then figure out how to get them there.
+
+Before you start sending messages, make sure you**_ _****know what you need to send over the wire** (and what you don’t). For example, in an action game, your player’s movements will probably be sent to the server many times per second, while your player’s inventory status only needs to be sent when it changes. Smaller data structures are less likely to pay fragmentation penalties. The fastest packet is the one you never send.
```

Comment:
I think an asterisk typo here unless I'm missing some special sequence.

Reply:
Almost definitely an error, it seems the Google Docs -> Markdown generator I used has some interesting asterisk quirks. Thanks for catching all of these @lugehorsam!

Reply:
```suggestion
Before you start sending messages, make sure you **know what you need to send over the wire** (and what you don’t). For example, in an action game, your player’s movements will probably be sent to the server many times per second, while your player’s inventory status only needs to be sent when it changes. Smaller data structures are less likely to pay fragmentation penalties. The fastest packet is the one you never send.
```